File: 0077-combinations/0077-combinations.cpp
class Solution {
public:
    void backtrack(vector<vector<int>>&ans, vector<int>list, int n, int k, int index){
        if(list.size()==k){
            ans.push_back(list);
            return;
        }
        for(int i=index; i<=n; i++){
            list.push_back(i);
            backtrack(ans, list, n, k, i+1);
            list.pop_back();
        }
    }
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>>ans;
        vector<int>list;
        backtrack(ans, list, n, k, 1);
        return ans;
    }
};

File: 0077-combinations/0077-combinations.java
class Solution {
    List<List<Integer>>ans;
    void helper(int index, int n, int k, List<Integer>list){
        if(index>n){
            return;
        }
        if(list.size()==k){
            ans.add(new ArrayList<>(list));
            return;
        }
        for(int i=index; i<n; i++){
            list.add(i+1);
            helper(i+1, n, k, list);
            list.remove(list.size()-1);
        }
    }
    public List<List<Integer>> combine(int n, int k) {
        ans = new ArrayList<>();
        helper(0, n, k, new ArrayList<>());
        return ans;
    }
}